### PR TITLE
[BEAM-2724] Support for IOTargetName like in updated IO dataflow metrics

### PR DESCRIPTION
**Pull request by pabloem**

### CODE REVIEW

**Comment by pabloem** [comment]
r: @aaltay 
This adds an IOTargetName attribute for internal counters. This will be used to measure time spent reading from side inputs and shuffle.

File: sdks/python/apache_beam/utils/counters_test.py [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+
+class CounterNameTest(unittest.TestCase):
+
+  def test_find_counter(self):
+    c1 = object()
+    cn1 = CounterName('cname1', 'stage1', 'step1')
+    c2 = object()
+    cn2 = CounterName('cname2', 'stage2', 'step2')
+
+    index = {cn1: c1, cn2: c2}
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/counters_test.py]: What is this testing? That CounterName is hashable?

File: sdks/python/apache_beam/utils/counters_test.py [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+    self.assertEqual(c1, index[cn1])
+    self.assertEqual(c2, index[cn2])
+
+  def test_iotarget_counter_names(self):
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/counters_test.py]: You combine all these cases into a single method.These are all equality tests.

File: sdks/python/apache_beam/utils/counters.py [hunk 3]
```diff
@@ -24,39 +24,84 @@
 """
 
 import threading
+
 from apache_beam.transforms import cy_combiners
 
 
+class IOTargetName(object):
+
+  def __init__(self,
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/counters.py]: Do you expect users to use this `__init__` method directly?

Perhaps you can use a `namedtuple` and move static methods out of it.

Comment (pabloem) [inline sdks/python/apache_beam/utils/counters.py]: Done. What do you think?

**Comment by pabloem** [comment]
Addressed comments.

File: sdks/python/apache_beam/utils/counters.py [hunk 4]
```diff
@@ -23,40 +23,69 @@
...
+  """Create an IOTargetName that identifies the reading of a side input."""
+  return IOTargetName(step_name, input_index, None)
+
+
+def shuffle_id(step_name):
+  """Create an IOTargetName that identifies a GBK step."""
+  return IOTargetName(None, None, step_name)
+
+
 class CounterName(object):
```
Comment (aaltay) [inline sdks/python/apache_beam/utils/counters.py]: Would it make sense to make this a namedtuple as well?

Comment (pabloem) [inline sdks/python/apache_beam/utils/counters.py]: YEs, it does. Updated to subclass namedtuple (to use default values and custom printing functions)

**Comment by aaltay** [comment]
LGTM. Thank you, I will merge after tests pass.

**Comment by aaltay** [comment]
@pabloem There is a lint error, please fix that.

**Comment by pabloem** [comment]
ugh sorry about that. I ran `run_pylint`, and that did not find the issue. Had to run it with `tox`. Fixed.

**Comment by aaltay** [comment]
`run_pylint` will run with whatever version of those tools you have installed, `tox -e lint` also controls the versions of those tools. It is possible that you have an old version with less strict rules.

**Comment by aaltay** [comment]
retest this please

**Comment by pabloem** [comment]
The tests passed after my latest push. See: https://builds.apache.org/job/beam_PreCommit_Java_MavenInstall/13981/

**Pull request closed without merging**